### PR TITLE
Msgpack format

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add MessagePack Input/Output format.

Detailed description / Documentation draft:

### **MsgPack Input/Output format**

Reading and writing MessagePack data files.

**Data types matching:**
| ClickHouse data type | MsgPack data type|
|:----------------|:---------|
| `UIntN` | `uint N`, `positive fixint` |
| `IntN` | `int N` |
| `String`, `FixedString` | `fixstr`, `str 8`, `str 16`, `str 32 `|
|`Float32`| `float 32`|
| `Float64` | `float 64`|
| `Date` | `uint 16` |
| `DateTime` | `uint 32` |
| `DateTime64` | `uint 64`|
| `Array` | `fixarray`, `array 16`, `array 32` |
| `Nothing` | `nil` |

### CODE REVIEW

File: dbms/src/Processors/Formats/Impl/MsgPackRowInputFormat.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            throw Exception("Not enough values to complete the row.", ErrorCodes::INCORRECT_DATA);
+        return false;
+    }
+    return true;
+}
+
+void registerInputFormatProcessorMsgPack(FormatFactory & factory)
+{
+    factory.registerInputFormatProcessor("MsgPack", [](
+            ReadBuffer &buf,
```
Comment (alexey-milovidov) [inline dbms/src/Processors/Formats/Impl/MsgPackRowInputFormat.cpp]: (minor) let's write whitespaces on both sides of &.

File: dbms/tests/queries/0_stateless/01098_msgpack_format.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. $CURDIR/../shell_config.sh
+
+$CLICKHOUSE_CLIENT --query="DROP TABLE IF EXISTS msgpack";
+$CLICKHOUSE_CLIENT --query="CREATE TABLE msgpack (uint8 UInt8, uint16 UInt16, uint32 UInt32, uint64 UInt64, int8 Int8, int16 Int16, int32 Int32, int64 Int64, float Float32, double Float64, string String, date Date, datetime DateTime, datetime64 DateTime64, array Array(UInt32)) ENGINE = Memory";
+
+
+$CLICKHOUSE_CLIENT --query="INSERT INTO msgpack VALUES (255, 65535, 4294967295, 100000000000, -128, -32768, -2147483648, -100000000000, 2.02, 10000.0000001, 'String', 18980, 1639872000, 1639872000000, [1,2,3,4,5]), (4, 1234, 3244467295, 500000000000, -1, -256, -14741221, -7000000000, 100.1, 14321.032141201, 'Another string', 20000, 1839882000, 1639872891123, [5,4,3,2,1]),(42, 42, 42, 42, 42, 42, 42, 42, 42.42, 42.42, '42', 42, 42, 42, [42])";
+
+$CLICKHOUSE_CLIENT --query="SELECT * FROM msgpack FORMAT MsgPack" > $CURDIR/data_msgpack/all_types.msgpk;
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/01098_msgpack_format.sh:12]: But if we generate this file, why we have it in git repo?

File: dbms/tests/queries/0_stateless/01098_msgpack_format.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/01098_msgpack_format.sh:1]: Do we have a test case for transparent type conversions while reading?
Like - write Array(UInt8) to msgpack, then read as Array(Int64)?

**Comment by Avogar** [comment]
I found an error in large test, I will fix it

**Comment by alexey-milovidov** [comment]
There is no way to disable Msgpack from build, but we will fix it later.

**Review by alexey-milovidov — APPROVED** [review]

File: cmake/find/msgpack.cmake [hunk 4]
```diff
@@ -0,0 +1,2 @@
+set(MSGPACK_INCLUDE_DIR ${ClickHouse_SOURCE_DIR}/contrib/msgpack-c/include)
+message(STATUS "Using msgpack: ${MSGPACK_INCLUDE_DIR}")
```
Comment (azat) [inline cmake/find/msgpack.cmake:2]: I guess that the same warning as other `find/*.cmake` has will be useful:

```cmake
message(WARNING "submodule contrib/msgpack is missing. to fix try run: \n git submodule update --init --recursive")
```

And what it is more useful (although for 0.001%) - add ability to use unbundled msgpack (at least almost all other contrib stuff allows this)

Comment (alexey-milovidov) [reply cmake/find/msgpack.cmake:2]: Yes. If you can quickly add it, it will be very useful.

**Comment by alexey-milovidov** [comment]
It does not work at all:

```
SELECT * FROM test.hits FORMAT MsgPack
```

```
Program received signal SIGSEGV, Segmentation fault.
memcpy (destination=0x7ffff5fb49e3, source=0x10, size=size@entry=1048541) at /usr/lib/gcc/x86_64-linux-gnu/9/include/emmintrin.h:727
727       *__P = __B;
(gdb) bt
#0  memcpy (destination=0x7ffff5fb49e3, source=0x10, size=size@entry=1048541) at /usr/lib/gcc/x86_64-linux-gnu/9/include/emmintrin.h:727
#1  0x0000000008fd1307 in memcpy (__len=<optimized out>, __src=<optimized out>, __dest=<optimized out>) at ../contrib/libc-headers/x86_64-linux-gnu/bits/string_fortified.h:34
#2  DB::WriteBuffer::write (this=0x7fffffffde00, from=0x10 <error: Cannot access memory at address 0x10>, n=4080715211) at ../src/IO/WriteBuffer.h:79
#3  0x000000000dbc8302 in msgpack::v1::packer<DB::WriteBuffer>::append_buffer<void, DB::WriteBuffer, unsigned long> (this=0x7ffff6c50e70, len=4080715211, 
    buf=0x10 <error: Cannot access memory at address 0x10>) at ../contrib/msgpack-c/include/msgpack/v1/pack.hpp:1255
#4  msgpack::v1::packer<DB::WriteBuffer>::append_buffer (len=4080715211, buf=0x10 <error: Cannot access memory at address 0x10>, this=0x7ffff6c50e70)
    at ../contrib/msgpack-c/include/msgpack/v1/pack.hpp:623
#5  msgpack::v1::packer<DB::WriteBuffer>::pack_str_body (l=4080715211, b=0x10 <error: Cannot access memory at address 0x10>, this=0x7ffff6c50e70) at ../contrib/msgpack-c/include/msgpack/v1/pack.hpp:1257
#6  DB::MsgPackRowOutputFormat::serializeField (this=this@entry=0x7ffff6c50d60, column=..., data_type=..., row_num=row_num@entry=0) at ../src/Processors/Formats/Impl/MsgPackRowOutputFormat.cpp:93
#7  0x000000000dbc892e in DB::MsgPackRowOutputFormat::write (this=<optimized out>, columns=..., row_num=0) at ../src/Common/COW.h:88
#8  0x000000000db3043a in DB::IRowOutputFormat::consume (this=this@entry=0x7ffff6c50d60, chunk=...) at ../src/Processors/Formats/IRowOutputFormat.cpp:26
#9  0x000000000db00199 in DB::IOutputFormat::write (block=..., this=0x7ffff6c50d60) at ../src/Processors/Formats/IOutputFormat.h:71
#10 DB::OutputStreamToOutputFormat::write (this=this@entry=0x7ffff64a7588, block=...) at ../src/Processors/Formats/OutputStreamToOutputFormat.cpp:14
#11 0x000000000da4ffc2 in DB::MaterializingBlockOutputStream::write (this=<optimized out>, block=...) at ../contrib/libcxx/include/memory:3826
#12 0x0000000009045ac1 in DB::Client::onData (block=..., this=<optimized out>) at ../programs/client/Client.cpp:1384
#13 DB::Client::onData (block=..., this=<optimized out>) at ../programs/client/Client.cpp:1370
#14 DB::Client::receiveAndProcessPacket (this=0x7fffffffca10) at ../programs/client/Client.cpp:1172
#15 0x0000000009046ac2 in DB::Client::receiveResult (this=0x7fffffffca10) at ../programs/client/Client.cpp:1154
#16 0x000000000904744a in DB::Client::processOrdinaryQuery (this=<optimized out>) at ../programs/client/Client.cpp:912
#17 DB::Client::processSingleQuery (this=<optimized out>, line=..., parsed_query_=...) at ../programs/client/Client.cpp:822
#18 0x000000000904829c in DB::Client::process (this=0x7fffffffca10, text=...) at ../contrib/libcxx/include/memory:3920
#19 0x00000000090489ed in DB::Client::nonInteractive (this=0x7fffffffca10) at ../programs/client/Client.cpp:653
#20 0x000000000904a044 in DB::Client::mainImpl (this=0x7fffffffca10) at ../programs/client/Client.cpp:560
#21 0x000000000904b846 in DB::Client::main (this=0x7fffffffca10) at ../programs/client/Client.cpp:235
#22 0x000000000e38bb13 in Poco::Util::Application::run (this=0x7fffffffca10) at ../contrib/poco/Util/src/Application.cpp:334
#23 0x000000000901f3c9 in mainEntryClickHouseClient (argc=3, argv=0x7ffff6c63c40) at ../programs/client/Client.cpp:1891
#24 0x0000000008f088e9 in main (argc_=<optimized out>, argv_=<optimized out>) at ../contrib/libcxx/include/vector:655
```

**Comment by adevyatova** [comment]
Internal documentation ticket DOCSUP-14372

**Pull request merged**